src/main/GCNodeAllocator.cpp

```diff
@@ -0,0 +1,499 @@
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+#include "rho/AddressSanitizer.hpp"
+#include "rho/AllocationTable.hpp"
+#include "rho/GCNodeAllocator.hpp"
+
+#ifdef HAVE_ADDRESS_SANITIZER
+// Quarantine free lists are used to store freed objects for a while before
+// they can be reused. Allocations are poisoned while in the quarantine. This
+// makes it more likely that Address Sanitizer can detect use-after-free
+// errors.
+rho::FreeListNode* rho::GCNodeAllocator::s_quarantine[s_num_freelists];
+#endif
+
+// The allocation table tracks medium and large allocations.
+rho::AllocationTable* rho::GCNodeAllocator::s_alloctable = nullptr;
+
+// Superblock pools by size class. Each superblock in linked in this
+// array has at least one untouched block available.
+rho::AllocatorSuperblock* rho::GCNodeAllocator::s_superblocks[
+    s_num_small_pools + s_num_medium_pools];
+
+// Tracking heap bounds for fast rejection in pointer lookup.
+uintptr_t rho::GCNodeAllocator::s_heap_start = UINTPTR_MAX;
+uintptr_t rho::GCNodeAllocator::s_heap_end = 0;
+
+// Free lists head pointers.
+rho::FreeListNode* rho::GCNodeAllocator::s_freelists[s_num_freelists];
+
+#ifdef ALLOCATION_CHECK
+// Helper function for allocator consistency checking.
+// An additional allocation map is added which shadows the state of the
+// allocator. The extra allocation map is checked to verify each operation
+// on the allocator.
+
+static void add_to_allocation_map(void* allocation, std::size_t size);
+static void remove_from_allocation_map(void* allocation);
+
+/**
+ * This should always return the same result as
+ * GCNodeAllocator::lookupPointer().  If it does not, something is wrong.
+ *
+ * When allocation checking is enabled, each call to lookupPointer() is checked
+ * against the result of this function.  This is also used to test if we try to
+ * free something unalloated or allocate something which is already allocated.
+ */
+static void* lookup_in_allocation_map(void* tentative_pointer);
+#endif
+
+void allocerr(const char* message) {
+  fprintf(stderr, "ERROR: %s\n", message);
+  abort();
+}
+
+/**
+ * Helper function for finding the the 2-log of the next power of two greater
+ * than or equal to size, such that (1 << result) >= size.
+ *
+ * Handles 32-bit integers.
+ */
+static unsigned next_log2_32(unsigned size) {
+  if (size == 0) {
+    return 0;
+  }
+
+  // Feature test macro for GCC builtins. Also available in recent Clangs.
+#ifndef __has_builtin
+#define __has_builtin(x) 0
+#endif
+
+  // First we figure out the 2-log of the highest set bit.
+  // Use builtin to count leading zeroes if available.
+#if __has_builtin(__builtin_clz)
+  unsigned log2 = 31 - __builtin_clz(size);
+#else
+  // Use manual log-2 finding implementation. The builtin version is available
+  // for most compilers, but if not we fall back on this version.
```

Comment:
If you adopted this algorithm from a third-party source then it may be a good idea to cite it here.